### ISSUE TITLE
GenieX can support Snapdragon X Plus 10-Core

### ISSUE THREAD

**Issue by Tom-TED** [qualcomm/GenieX#1153]
C:\Users\iis2g\AppData\Local\GenieX CLI>geniex infer ai-hub-models/Qwen3-VL-4B-Instruct
Model is not currently cached, downloading...
Error: download model failed: SDKError(Unknown error code), chipset "Snapdragon X Plus 8-Core CRD" not available for this model; supported: Dragonwing IQ-8275 EVK, Dragonwing IQ-9075 EVK, SA8775P ADP, Snapdragon 8 Elite Gen 5 QRD, Snapdragon 8 Elite QRD, Snapdragon X Elite CRD, Snapdragon X2 Elite CRD

I'm using Copilot+PC with Snapdragon X Plus 10-Core but I encountered the above error. I believe 10-Core Plus chipset is similar to Elite. Do you have a plan to support Plus 10-Core ? or how to use Plus 10-Core to run GenieX ?

**Comment by zhiyuan8** [comment]
Snapdragon X plus has same NPU dsp arch as X Elite, `v73`, so it should support X Plus. We will examine the `geniex pull` artifacts path on qualcomm ai hub and get back to you.

**Comment by flatlines** [comment]
+1

I can confirm that this is still an issue with GenieX CLI Version: v0.3.16:

> Error: download model failed: SDKError(Unknown error code), chipset "Snapdragon X Plus 8-Core CRD" not available for this model; supported: Dragonwing IQ-9075 EVK, SA8775P ADP, Snapdragon 8 Elite Gen 5 QRD, Snapdragon 8 Elite QRD, Snapdragon X Elite CRD, Snapdragon X2 Elite CRD

My chipset is the Snapdragon(R) X 10-core X1P64100 @ 3.40 GHz (3.42 GHz) processor on Windows 11.

The team over at qualcomm ai hub suggested I open a feature request here.

Thanks for the work you are all doing on the CLI, I am really stoked to try out these QAIRT models and see how they run!

**Comment by flatlines** [comment]
I wanted to provide a quick update on this. The latest version of GenieX CLI:

```
GenieX CLI Version:     v0.5.0
QAIRT Runtime Version:  v2.45.0.260326
LlamaCPP Runtime Hash:  873e5d8
```
Now runs QAIRT models on the Snapdragon(R) X 10-core X1P64100 @ 3.40 GHz (3.42 GHz) on Windows. However, the QAIRT builds seem to run more slowly than the llama.cpp builds. I think this may be related to the following DCP errors that appear anytime a QAIRT model is run:
```
DSP_INFO UNSUPPORTED_KEY: 49
DSP_INFO UNSUPPORTED_KEY: 50
DSP_INFO UNSUPPORTED_KEY: 49
DSP_INFO UNSUPPORTED_KEY: 50
DSP_INFO UNSUPPORTED_KEY: 49
DSP_INFO UNSUPPORTED_KEY: 50
```
The following command `geniex config list` returns `chipset: Snapdragon X Elite CRD`.

**Comment by Tom-TED** [comment]
I tried to use v0.5.0 and it's working fine with gemma-4-E4B.
Performance : 13.8 tok/s • 966 tok • 0.3 s first token

C:\Users\iis2g\AppData\Local\GenieX CLI>geniex version
GenieX CLI Version:     v0.3.19
QAIRT Runtime Version:  v2.45.0.260326
LlamaCPP Runtime Hash:  6ba5ef2

C:\Users\iis2g\AppData\Local\GenieX CLI>geniex infer google/gemma-4-E4B-it-qat-q4_0-gguf
> what is Qualcomm QNN?
**Qualcomm QNN** (or more accurately, **Qualcomm Neural Network SDK**) is a set of tools and a runtime environment provided by Qualcomm that allows developers to **optimize, deploy, and run machine learning (ML) models efficiently on Qualcomm Snapdragon mobile processors.**

In essence, it is a framework designed to bridge the gap between a trained ML model (like one created in TensorFlow, PyTorch, or Core ML) and the actual hardware acceleration capabilities of a mobile device.

Here is a detailed breakdown of what QNN is, what it does, and why it is important:

---

### 🧠 What Does QNN Do? (The Core Function)

The primary goal of QNN is **inference acceleration** on edge devices (like smartphones). Machine learning models, especially deep neural networks (DNN), require massive amounts of computation. Running these computations efficiently on a mobile CPU or GPU can be slow, drain the battery quickly, and generate excessive heat.

QNN solves this by:

1. **Model Optimization:** It takes a trained ML model (which might be in a standard format like ONNX) and analyzes its computational graph.
2. **Hardware Mapping:** It determines which parts of the model can be accelerated by specific hardware units within the Snapdragon chip (e.g., the DSP, the dedicated NPU/AI engine, or the GPU).
3. **Runtime Execution:** It compiles and executes the model using a highly optimized runtime that leverages these specialized hardware blocks, resulting in **significantly faster inference times** and **lower power consumption** compared to running the model on the general-purpose CPU.

### 🛠️ Key Components and Concepts

* **Inference Engine:** QNN acts as the execution engine. It doesn't train the models; it only runs the *already trained* models (inference).
* **Delegates/Accelerators:** Within the QNN framework, there are components that act as "delegates." A developer can tell QNN, "Run this layer on the DSP," or "Run this matrix multiplication on the NPU."
* **Optimization:** It handles quantization (reducing the precision of model weights, e.g., from 32-bit floats to 8-bit integers) to make the model smaller and faster for mobile execution without significant accuracy loss.
* **Integration:** QNN integrates deeply with the Android ecosystem and the specific hardware architecture of Qualcomm chips.

### 💡 Why is QNN Important? (The "Why")

In the world of mobile AI, speed and power efficiency are everything. QNN is critical because it allows developers to:

1. **Enable Real-Time Features:** Features like real-time object detection, advanced background blurring, complex voice recognition, and on-device language translation require extremely low latency. QNN makes this possible.
2. **Save Battery Life:** By offloading heavy computations from the power-hungry CPU to specialized, energy-efficient accelerators (like the DSP/NPU), QNN drastically reduces power draw.
3. **Ensure Privacy:** Because the model runs entirely *on the device* (on-device inference), sensitive data (like photos or voice recordings) never needs to be sent to the cloud, enhancing user privacy.

### 🆚 QNN vs. Other Framework Components

It's helpful to distinguish QNN from other ML components:

| Component | Role | Function |
| :--- | :--- | :--- |
| **TensorFlow/PyTorch** | **Training Framework** | Used by data scientists to *create* and *train* the model using vast datasets. |
| **ONNX** | **Interoperability Standard** | A common intermediate format used to export models from PyTorch/TF so they can be used by multiple runtimes. |
| **Qualcomm QNN** | **Inference Engine/Runtime** | Takes the ONNX/model file and *optimizes and executes* it specifically for the Snapdragon chip's hardware. |
| **Snapdragon NPU/DSP** | **Hardware Accelerator** | The physical silicon blocks that perform the actual math very quickly and efficiently. |

### 🚀 Summary Analogy

Think of it this way:

* **Training Framework (PyTorch/TF):** The Architect who designs the blueprint of a building.
* **ONNX:** A standardized blueprint format.
* **Qualcomm QNN:** The specialized Construction Manager who reads the blueprint and knows exactly which specialized construction crews (DSP, NPU) to deploy to build the structure as fast and efficiently as possible on the specific piece of land (the Snapdragon chip).
* **Snapdragon Hardware:** The construction crews themselves.

— 13.8 tok/s • 966 tok • 0.3 s first token —
